sdks/python/apache_beam/runners/common.py

```diff
@@ -51,117 +51,192 @@ def receive(self, windowed_value):
+                    the method.
 
-    if logging_context:
-      self.logging_context = logging_context
-    else:
-      self.logging_context = get_logging_context(logger, step_name=step_name)
+    Using a regular method call here instead of __call__ to reduce per-element
+    overhead.
+    """
+    return self._method_value(*input_args, **input_kwargs)
 
-    # Optimize for the common case.
-    self.main_receivers = as_receiver(tagged_receivers[None])
 
-    # TODO(sourabh): Deprecate the use of context
-    if state:
-      assert context is None
-      self.context = DoFnContext(self.step_name, state=state)
+class DoFnSignature(object):
+  """Represents the signature of a given ``DoFn`` object.
+
+  Signature of a ``DoFn`` provides a view of the properties of a given ``DoFn``.
+  Among other things, this will give an extensible way for for (1) accessing the
+  structure of the ``DoFn`` including methods and method parameters
+  (2) identifying features that a given ``DoFn`` support, for example, whether
+  a given ``DoFn`` is a Splittable ``DoFn`` (
+  https://s.apache.org/splittable-do-fn) (3) validating a ``DoFn`` based on the
+  feature set offered by it.
+  """
+
+  def __init__(self, do_fn):
+    # We add a property here for all methods defined by Beam DoFn features.
+    self.process_method = None
+    self.start_bundle_method = None
+    self.finish_bundle_method = None
+
+    assert isinstance(do_fn, core.DoFn)
+    self.do_fn = do_fn
+
+    def _create_do_fn_method(do_fn, method_name):
+      arguments, _, _, defaults = do_fn.get_function_arguments(method_name)
+      defaults = defaults if defaults else []
+      method_value = getattr(do_fn, method_name)
+      return DoFnMethodWrapper(method_value, arguments, defaults)
+
+    self.process_method = _create_do_fn_method(do_fn, 'process')
+    self.start_bundle_method = _create_do_fn_method(do_fn, 'start_bundle')
+    self.finish_bundle_method = _create_do_fn_method(do_fn, 'finish_bundle')
+    self._validate()
+
+  def _validate(self):
+    # start_bundle and finish_bundle methods should only have ContextParam as a
+    # default argument.
+    self._validate_start_bundle()
+    self._validate_finish_bundle()
+
+  def _validate_start_bundle(self):
+    self._validate_bundle_method()
+
+  def _validate_finish_bundle(self):
+    self._validate_bundle_method()
+
+  def _validate_bundle_method(self):
+    assert core.DoFn.ElementParam not in self.start_bundle_method.defaults
+    assert core.DoFn.SideInputParam not in self.start_bundle_method.defaults
+    assert core.DoFn.TimestampParam not in self.start_bundle_method.defaults
+    assert core.DoFn.WindowParam not in self.start_bundle_method.defaults
+
+
+class DoFnInvoker(object):
+  """An abstraction that can be used to execute DoFn methods.
+
+  A DoFnInvoker describes a particular way for invoking methods of a DoFn
+  represented by a given DoFnSignature."""
+
+  def __init__(self, signature):
```

Comment:
Should keep the constructor across all invoker implementations and then each subclass can choose to use it or not ??